### DISCUSSION TITLE
How to solve teams load limit in bot?

### DISCUSSION THREAD

**Discussion by hueiyuan** [microsoft/teams-sdk#528] [Q&A]
We currently plan to use azure openai model to implement retrieval document bot. We found that once bot agent have many chain and tools, the response will take 30 - 60 seconds to reply user. And MS Teams app have 30 seconds limitation (https://learn.microsoft.com/en-us/power-apps/teams/known-issues-limitations#load-time-limit). 

Have any ideas can improve it or solve? 
If you are willing to share your solution and tips, I will be appreciated it, thank you

**Comment by corinagum** [comment]
Hi @hueiyuan, I'm asking around to see if there devs with more experience on load time limit to help with your question. 

Could you summarize how many chains / tools your app has, to get an idea of where you start to have delays? If you start removing some of the tooling, at what point (how many) items are there when the delay starts to be noticeable?

[edit]
Also, the documentation you linked to is regarding power apps; are you using a power app with Azure Open AI?

**Reply by corinagum** [reply]
Hi @hueiyuan, in a meeting yesterday discussing our upcoming changes before GA, we learned that `chain` will be replaced with a much more efficient system. So while they will be breaking changes, I anticipate they will lead to a much more performant app. My suggestion would be to pause while we wait for those changes. Thanks for your patience!

**Reply by hueiyuan** [reply]
@corinagum 
Thanks for your assistance. 
Currently we find the workaround to solve it. Just using PVA + automate, and using HTTP action to call our endpoint.
And based our prompt, tool to request openai to get response.

So based on your suggestion, if we would like to use `teams-ai`. we need to wait your changes, right?

**Reply by corinagum** [reply]
@hueiyuan that's great that you found the workaround -- it would be super appreciated if you share your code here for that workaround. 
To clarify, are you using teams-ai already? From the comments about chains, it sounds like you are already using our library. If things are working for you with your chain workaround, there's no issues with using teams-ai. If you are not using it already, though due to my lack of experience with PVA, I'm not familiar with how that migration process will work. Let me know what your project's current status is :)

In terms of migrating away from chain, yes; that work will be a few weeks' wait. Are you writing in JS? C# and Python will follow JS by a cycle or two for porting. But again, it is not a blocker for migrating to teams-ai if you have a workaround.
